File: clevertap-core/src/main/java/com/clevertap/android/sdk/ActivityLifeCycleManager.java
package com.clevertap.android.sdk;

import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.os.RemoteException;
import com.android.installreferrer.api.InstallReferrerClient;
import com.android.installreferrer.api.InstallReferrerStateListener;
import com.android.installreferrer.api.ReferrerDetails;
import com.clevertap.android.sdk.events.BaseEventQueueManager;
import com.clevertap.android.sdk.inapp.InAppController;
import com.clevertap.android.sdk.pushnotification.PushProviders;
import com.clevertap.android.sdk.task.CTExecutorFactory;
import com.clevertap.android.sdk.task.Task;
import java.util.concurrent.Callable;

class ActivityLifeCycleManager {

    private final AnalyticsManager analyticsManager;

    private final BaseEventQueueManager baseEventQueueManager;

    private final BaseCallbackManager callbackManager;

    private final CleverTapInstanceConfig config;

    private final Context context;

    private final CoreMetaData coreMetaData;

    private final InAppController inAppController;

    private final PushProviders pushProviders;

    private final SessionManager sessionManager;

    ActivityLifeCycleManager(Context context,
            CleverTapInstanceConfig config,
            AnalyticsManager analyticsManager,
            CoreMetaData coreMetaData,
            SessionManager sessionManager,
            PushProviders pushProviders,
            BaseCallbackManager callbackManager,
            InAppController inAppController,
            BaseEventQueueManager baseEventQueueManager) {
        this.context = context;
        this.config = config;
        this.analyticsManager = analyticsManager;
        this.coreMetaData = coreMetaData;
        this.sessionManager = sessionManager;
        this.pushProviders = pushProviders;
        this.callbackManager = callbackManager;
        this.inAppController = inAppController;
        this.baseEventQueueManager = baseEventQueueManager;
    }

    //Lifecycle
    public void activityPaused() {
        CoreMetaData.setAppForeground(false);
        sessionManager.setAppLastSeen(System.currentTimeMillis());
        config.getLogger().verbose(config.getAccountId(), "App in background");
        final int now = (int) (System.currentTimeMillis() / 1000);
        if (coreMetaData.inCurrentSession()) {
            try {
                StorageHelper
                        .putInt(context,
                                StorageHelper.storageKeyWithSuffix(config, Constants.LAST_SESSION_EPOCH),
                                now);
                config.getLogger().verbose(config.getAccountId(), "Updated session time: " + now);
            } catch (Throwable t) {
                config.getLogger()
                        .verbose(config.getAccountId(), "Failed to update session time time: " + t.getMessage());
            }
        }
    }

    //Lifecycle
    public void activityResumed(Activity activity) {
        config.getLogger().verbose(config.getAccountId(), "App in foreground");
        sessionManager.checkTimeoutSession();
        //Anything in this If block will run once per App Launch.
        //Will not run for Apps which disable App Launched event
        if (!coreMetaData.isAppLaunchPushed()) {

            analyticsManager.pushAppLaunchedEvent();
            analyticsManager.fetchFeatureFlags();
            pushProviders.onTokenRefresh();
            Task<Void> task = CTExecutorFactory.executors(config).postAsyncSafelyTask();
            task.execute("HandlingInstallReferrer",new Callable<Void>() {
                @Override
                public Void call() {
                    if (!coreMetaData.isInstallReferrerDataSent() && coreMetaData
                            .isFirstSession()) {
                        handleInstallReferrerOnFirstInstall();
                    }
                    return null;
                }
            });

            try {
                if (callbackManager.getGeofenceCallback() != null) {
                    callbackManager.getGeofenceCallback().triggerLocation();
                }
            } catch (IllegalStateException e) {
                config.getLogger().verbose(config.getAccountId(), e.getLocalizedMessage());
            } catch (Exception e) {
                config.getLogger().verbose(config.getAccountId(), "Failed to trigger location");
            }
        }
        baseEventQueueManager.pushInitialEventsAsync();
        inAppController.checkExistingInAppNotifications(activity);
        inAppController.checkPendingInAppNotifications(activity);
    }

    public void onActivityCreated(final Bundle notification, final Uri deepLink) {
        try {
            boolean shouldProcess = config.isDefaultInstance();

            if (shouldProcess) {
                if (notification != null && !notification.isEmpty() && notification
                        .containsKey(Constants.NOTIFICATION_TAG)) {
                    analyticsManager.pushNotificationClickedEvent(notification);
                }

                if (deepLink != null) {
                    try {
                        analyticsManager.pushDeepLink(deepLink, false);
                    } catch (Throwable t) {
                        // no-op
                    }
                }
            }
        } catch (Throwable t) {
            Logger.v("Throwable - " + t.getLocalizedMessage());
        }
    }

    private void handleInstallReferrerOnFirstInstall() {
        config.getLogger().verbose(config.getAccountId(), "Starting to handle install referrer");
        try {
            final InstallReferrerClient referrerClient = InstallReferrerClient.newBuilder(context).build();
            referrerClient.startConnection(new InstallReferrerStateListener() {
                @Override
                public void onInstallReferrerServiceDisconnected() {
                    if (!coreMetaData.isInstallReferrerDataSent()) {
                        handleInstallReferrerOnFirstInstall();
                    }
                }

                @Override
                public void onInstallReferrerSetupFinished(int responseCode) {
                    switch (responseCode) {
                        case InstallReferrerClient.InstallReferrerResponse.OK:
                            // Connection established.
                            ReferrerDetails response;
                            try {
                                response = referrerClient.getInstallReferrer();
                                String referrerUrl = response.getInstallReferrer();
                                coreMetaData
                                        .setReferrerClickTime(response.getReferrerClickTimestampSeconds());
                                coreMetaData
                                        .setAppInstallTime(response.getInstallBeginTimestampSeconds());
                                analyticsManager.pushInstallReferrer(referrerUrl);
                                coreMetaData.setInstallReferrerDataSent(true);
                                config.getLogger().debug(config.getAccountId(),
                                        "Install Referrer data set [Referrer URL-" + referrerUrl + "]");
                            } catch (RemoteException e) {
                                config.getLogger().debug(config.getAccountId(),
                                        "Remote exception caused by Google Play Install Referrer library - " + e
                                                .getMessage());
                                referrerClient.endConnection();
                                coreMetaData.setInstallReferrerDataSent(false);
                            }catch (NullPointerException npe){
                                config.getLogger().debug(config.getAccountId(),
                                        "Install referrer client null pointer exception caused by Google Play Install Referrer library - " + npe
                                                .getMessage());
                                referrerClient.endConnection();
                                coreMetaData.setInstallReferrerDataSent(false);
                            }
                            break;
                        case InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
                            // API not available on the current Play Store app.
                            config.getLogger().debug(config.getAccountId(),
                                    "Install Referrer data not set, API not supported by Play Store on device");
                            break;
                        case InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE:
                            // Connection couldn't be established.
                            config.getLogger().debug(config.getAccountId(),
                                    "Install Referrer data not set, connection to Play Store unavailable");
                            break;
                    }
                }
            });
        } catch (Throwable t) {
            config.getLogger().verbose(config.getAccountId(),
                    "Google Play Install Referrer's InstallReferrerClient Class not found - " + t
                            .getLocalizedMessage()
                            + " \n Please add implementation 'com.android.installreferrer:installreferrer:2.1' to your build.gradle");
        }
    }
}

File: clevertap-core/src/main/java/com/clevertap/android/sdk/ActivityLifeCycleManager.kt
package com.clevertap.android.sdk

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.os.RemoteException
import androidx.annotation.RestrictTo
import com.android.installreferrer.api.InstallReferrerClient
import com.android.installreferrer.api.InstallReferrerStateListener
import com.android.installreferrer.api.ReferrerDetails
import com.clevertap.android.sdk.StorageHelper.putInt
import com.clevertap.android.sdk.events.BaseEventQueueManager
import com.clevertap.android.sdk.inapp.InAppController
import com.clevertap.android.sdk.pushnotification.PushProviders
import com.clevertap.android.sdk.task.CTExecutors
import com.clevertap.android.sdk.utils.Clock

/**
 * Manages the activity lifecycle events for the CleverTap SDK.
 *
 * This class is responsible for handling various lifecycle callbacks of an Android application,
 * such as when an activity is created, resumed, or paused. It orchestrates actions like
 * session management, pushing "App Launched" events, handling install referrers,
 * processing deep links and push notification clicks, and triggering in-app messages.
 *
 * This is an internal class and is not meant to be used directly by developers.
 *
 * @property context The application context.
 * @property config The instance-specific configuration.
 * @property analyticsManager The manager for handling analytics events.
 * @property coreMetaData The manager for core metadata like session and device info.
 * @property sessionManager The manager for user sessions.
 * @property pushProviders The manager for push notification services.
 * @property callbackManager The manager for various SDK callbacks.
 * @property inAppController The controller for in-app notifications.
 * @property baseEventQueueManager The manager for the event queue.
 * @property executors The task executor for running background operations.
 * @property clock A clock instance for time-related operations.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
class ActivityLifeCycleManager internal constructor(
    private val context: Context,
    private val config: CleverTapInstanceConfig,
    private val analyticsManager: AnalyticsManager,
    private val coreMetaData: CoreMetaData,
    private val sessionManager: SessionManager,
    private val pushProviders: PushProviders,
    private val callbackManager: BaseCallbackManager,
    private val inAppController: InAppController,
    private val baseEventQueueManager: BaseEventQueueManager,
    private val executors: CTExecutors,
    private val clock: Clock
) {
    //Lifecycle
    fun activityPaused() {
        CoreMetaData.setAppForeground(false)
        sessionManager.appLastSeen = clock.currentTimeMillis()
        config.getLogger().verbose(config.accountId, "App in background")
        val task = executors.postAsyncSafelyTask<Unit>()
        task.execute(
            "activityPaused"
        ) {
            val now = clock.currentTimeSecondsInt()
            if (coreMetaData.inCurrentSession()) {
                try {
                    putInt(
                        context,
                        config.accountId,
                        Constants.LAST_SESSION_EPOCH,
                        now
                    )
                    config.getLogger()
                        .verbose(config.accountId, "Updated session time: $now")
                } catch (t: Throwable) {
                    config.getLogger().verbose(
                        config.accountId,
                        "Failed to update session time time: " + t.message
                    )
                }
            }
        }
    }

    //Lifecycle
    fun activityResumed(activity: Activity?) {
        config.getLogger().verbose(config.accountId, "App in foreground")
        sessionManager.checkTimeoutSession()

        //Anything in this If block will run once per App Launch.
        if (!coreMetaData.isAppLaunchPushed) {
            analyticsManager.pushAppLaunchedEvent()
            analyticsManager.fetchFeatureFlags()
            pushProviders.onTokenRefresh()
            val task = executors.postAsyncSafelyTask<Unit>()
            task.execute("HandlingInstallReferrer") {
                if (!coreMetaData.isInstallReferrerDataSent && coreMetaData
                        .isFirstSession
                ) {
                    handleInstallReferrerOnFirstInstall()
                }
            }

            val cleanUpTask = executors.ioTask<Unit>()
            cleanUpTask.execute("CleanUpOldGIFs") {
                Utils.cleanupOldGIFs(context, config, clock)
            }

            try {
                if (callbackManager.getGeofenceCallback() != null) {
                    callbackManager.getGeofenceCallback().triggerLocation()
                }
            } catch (e: IllegalStateException) {
                config.getLogger().verbose(config.accountId, e.localizedMessage)
            } catch (_: Exception) {
                config.getLogger().verbose(config.accountId, "Failed to trigger location")
            }
        }
        baseEventQueueManager.pushInitialEventsAsync()
        inAppController.showNotificationIfAvailable()
    }

    fun onActivityCreated(notification: Bundle?, deepLink: Uri?, accountId: String?) {
        try {
            val shouldProcess =
                (accountId == null && config.isDefaultInstance) || config.accountId == accountId

            if (shouldProcess) {
                if (notification != null && !notification.isEmpty() && notification
                        .containsKey(Constants.NOTIFICATION_TAG)
                ) {
                    analyticsManager.pushNotificationClickedEvent(notification)
                }

                if (deepLink != null) {
                    try {
                        analyticsManager.pushDeepLink(deepLink, false)
                    } catch (_: Throwable) {
                        // no-op
                    }
                }
            }
        } catch (t: Throwable) {
            Logger.v("Throwable - " + t.localizedMessage)
        }
    }

    private fun handleInstallReferrerOnFirstInstall() {
        config.getLogger().verbose(config.accountId, "Starting to handle install referrer")
        try {
            val referrerClient = InstallReferrerClient.newBuilder(context).build()
            referrerClient.startConnection(object : InstallReferrerStateListener {
                override fun onInstallReferrerServiceDisconnected() {
                    if (!coreMetaData.isInstallReferrerDataSent) {
                        handleInstallReferrerOnFirstInstall()
                    }
                }

                override fun onInstallReferrerSetupFinished(responseCode: Int) {
                    when (responseCode) {
                        InstallReferrerClient.InstallReferrerResponse.OK -> {
                            // Connection established
                            val task = executors.postAsyncSafelyTask<ReferrerDetails?>()

                            task.addOnSuccessListener { response: ReferrerDetails? ->
                                try {
                                    val referrerUrl = response!!.installReferrer
                                    coreMetaData.referrerClickTime =
                                        response.referrerClickTimestampSeconds
                                    coreMetaData.appInstallTime =
                                        response.installBeginTimestampSeconds
                                    analyticsManager.pushInstallReferrer(referrerUrl)
                                    coreMetaData.isInstallReferrerDataSent = true
                                    config.getLogger().debug(
                                        config.accountId,
                                        "Install Referrer data set [Referrer URL-$referrerUrl]"
                                    )
                                } catch (npe: NullPointerException) {
                                    config.getLogger().debug(
                                        config.accountId,
                                        "Install referrer client null pointer exception caused by Google Play Install Referrer library - "
                                                + npe
                                            .message
                                    )
                                    referrerClient.endConnection()
                                    coreMetaData.isInstallReferrerDataSent = false
                                }
                            }

                            task.execute("ActivityLifeCycleManager#getInstallReferrer") {
                                var response: ReferrerDetails? = null
                                try {
                                    response = referrerClient.installReferrer
                                } catch (e: RemoteException) {
                                    config.getLogger().debug(
                                        config.accountId,
                                        "Remote exception caused by Google Play Install Referrer library - " + e
                                            .message
                                    )
                                    referrerClient.endConnection()
                                    coreMetaData.isInstallReferrerDataSent = false
                                }
                                response
                            }
                        }

                        // API not available on the current Play Store app.
                        InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED ->
                            config.getLogger().debug(
                                config.accountId,
                                "Install Referrer data not set, API not supported by Play Store on device"
                            )

                        // Connection couldn't be established.
                        InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE ->
                            config.getLogger().debug(
                                config.accountId,
                                "Install Referrer data not set, connection to Play Store unavailable"
                            )
                    }
                }
            })
        } catch (t: Throwable) {
            config.getLogger().verbose(
                config.accountId,
                ("Google Play Install Referrer's InstallReferrerClient Class not found - " + t
                    .localizedMessage
                        + " \n Please add implementation 'com.android.installreferrer:installreferrer:2.1' to your build.gradle")
            )
        }
    }
}